File: 0410-split-array-largest-sum/0410-split-array-largest-sum.cpp
class Solution {
public:
    
    bool canSplit(vector<int>& nums, int k, long long maxSum) {
        long long currentSum = 0;
        int count = 1;  
        
        for(int num : nums){
            if (currentSum + num <= maxSum) {
                currentSum += num;
            }else{
                count++;
                currentSum = num;
                
                if(count > k){
                    return false;
                }
            }
        }
        
        return true;
    }
    
    int splitArray(vector<int>& nums, int k) {
        long long low = *max_element(nums.begin(), nums.end());
        long long high = accumulate(nums.begin(), nums.end(), 0LL);
        
        while(low <= high){
            long long mid = low + (high - low) / 2;
            
            if(canSplit(nums, k, mid)){
                high = mid - 1;  
            }else{
                low = mid + 1;   
            }
        }
        
        return low;
    }
};

File: 0410-split-array-largest-sum/0410-split-array-largest-sum.java
class Solution {
    public int splitArray(int[] nums, int k) {
        int maxEl = 0;
        int sumOEL = 0;
        for(int x : nums){
            maxEl = Math.max(maxEl,x);
            sumOEL += x;
        }
        int left = maxEl;
        int right = sumOEL;
        int ans = -1;
    
        while(left <= right){
            int mid = (left + right)/2;
            int partition = 1;
            int maxSum = mid;
            int CurrSum = 0;
            for( int el : nums){
                if(CurrSum + el <= maxSum){
                    CurrSum += el;
                }else{
                    partition++;
                    CurrSum = el;
                }
            }
            if(partition <= k){
                ans = maxSum;
                right = mid - 1;
            }else{
                left = mid + 1;
            }
        }
        return ans;
            
    }
}